### ISSUE TITLE
Add e-notice opt-in feature for fees communication

### ISSUE THREAD

**Issue by dougwaldron** [gaepdit/geco#483]
Requested by fees group, but could be extended to other groups as well.
[GECO E-Notice Opt-in.pdf](https://github.com/gaepdit/geco/files/6298789/GECO.E-Notice.Opt-in.pdf)

## GECO tasks

* [X] Remove contacts from main Facility page.
* [X] Add "Communication Preferences" tab.
* [X] Provide web page mock-ups:
    * [X] Communication Preferences summary page.
    * [X] Communication Preferences edit page.
    * [X] Mandatory confirmation page.
        - Passive confirmation option.
        - Active response required option.
* [x] #482 
* [x] Remove contacts from EIS pages in GECO
* [x] Create the Communication Preferences summary page.
* [x] Create the Communication Preferences edit page.
    - Save per-category preferences (email, electronic, or both).
    - Require one physical mail address.
    - Allow multiple optional email addresses.
    - Some categories don't use the electronic communication option.
* [x] Code the mandatory confirmation pages to be shown initially and periodically (annually).
* [x] Create an email verification page.
* [x] Draft verification email template.
* [x] Send to program for UAT.
* [x] Search for uses of other contact tables and address as appropriate.
    * [x] `APBContactInformation`
    * [x] `FS_CONTACTINFO`
    * [x] `FS_MAILOUT` 
* [x] Separate Name data into components parts.
    * [x] Update stored procedures
    * [x] Update GECO
* [x] Reconsider how to handle the communication confirmation page if some contact type addresses are missing.

### DB stored procedures

- [x] Summary page.
- [x] Edit page.
- [x] Confirmation.
- [x] Email verification
- [x] Migrate existing contacts based on how they were entered.

### Data migration

- [x] Clean up existing data.
- [x] Write data migration script.

### Mockups feedback

- [x] Add property to Category to enable/disable the electronic communication option.
- [x] Make telephone number required.

### UAT feedback

- [x] Extend email confirmation expiration to one week.

### Handle existing contacts tables

* [x] Update `APBContactInformation` whenever `Geco_MailContact` is updated in GECO.
* [x] Fix contact handling in:
    * [x] Annual Fees submission form
    * [x] Emissions Statement form

### Second UAT feedback

* [x] (Implied) Add email field to primary contact.
* [x] IAIP: Include email address column in fee mailout list table.
* [x] IAIP: Add communication preference column on fee statistics tables.
* [x] (Implied) Reconsider requiring confirmation of email addresses and/or require confirmation of primary contact email.
* [x] Require address confirmation/entry after selecting communication preference.
* [x] Future: How to proceed with automated emails/email templates?

### Third UAT feedback

* [x] IAIP: Add communication preference column on fee statistics *details* tables.

### Fourth UAT feedback

- [x] Program want to postpone preference selection and mandatory update screens until next year.

## IAIP

* [ ] gaepdit/iaip#1057

**Comment by dougwaldron** [comment]
### Project tasks

* [ ] Provide web page mock-ups
* [ ] Create new "Communication Preferences" tab on Facility page
* [ ] Create pages/panels to edit contacts/communication prefs
    - Include email confirmation steps
    - Draft email template
* [ ] Create mandatory confirmation page to be shown periodically
* [ ] Add communication preferences data layer
* [ ] Improve User Access tab for clarity
* [ ] Modify IAIP mailout list to include communication prefs/email confirmation status

**Comment by dougwaldron** [comment]
Notes:
[Scan May 21, 2021 at 2.32 PM.pdf](https://github.com/gaepdit/geco/files/6524662/Scan.May.21.2021.at.2.32.PM.pdf)
[Scan May 21, 2021 at 2.32 PM 1.pdf](https://github.com/gaepdit/geco/files/6524663/Scan.May.21.2021.at.2.32.PM.1.pdf)

**Comment by dougwaldron** [comment]
<details>
<summary>Sent mockups to program for review</summary>

---

These first few mockups show how the application will look on a standard basis.
The facility home page is simplified (contact editing has been removed from this page) and has a new “Communication Preference” tab.

![GECO Facility Summary](https://user-images.githubusercontent.com/737326/121424241-1fe29e00-c93f-11eb-823f-41898cc35782.png)

---

The communication preferences tab shows current settings. Preferences can be set per “category.” Mail contacts will be included for all current programs (Emission Inventory, Permitting, etc.), but we would need to discuss it with each program before including the option of electronic communications.
 
![GECO Facility Contacts(2)](https://user-images.githubusercontent.com/737326/121424286-2c66f680-c93f-11eb-9c13-20362c3ced80.png)

---

The edit button for each category takes the user to an edit page where preferences can be changed.

![GECO Edit Facility Contacts](https://user-images.githubusercontent.com/737326/121424331-38eb4f00-c93f-11eb-8e21-e1ed03369a31.png)

Upon adding a new email recipient, a verification email will be sent to that address with a link that the user has to click to confirm that they want to accept electronic communications.

---

The next couple of mockups show how the mandatory confirmation page might work. 

The first time a user goes to the facility page, they will be redirected to an announcement/confirmation page. The facility pages can’t be accessed until a selection is made.
 
![GECO Communication Preferences](https://user-images.githubusercontent.com/737326/121424399-486a9800-c93f-11eb-98be-d87ad465575a.png)

After submitting a preference, they will be shown the previous Edit page so that the mailing address and emails can be entered.

One possible drawback here is that we’re only requiring a response for fees, and other programs won’t get a similar page (we can’t assume a facility would make the same choice for each category). This isn’t necessarily a bad thing because the user will still be referred to the communication preferences page, and so far only fees has made this a priority.

---

Finally, once an initial response has been submitted, users can routinely (annually?) be shown a reminder page asking them to review & confirm their current settings.

![GECO Facility Contacts(1)](https://user-images.githubusercontent.com/737326/121424462-5cae9500-c93f-11eb-9e4b-59fbbbbfd032.png)

 </details>

**Comment by dougwaldron** [comment]
Note for migrating dates:
https://github.com/gaepdit/enforcement-orders/issues/10#issuecomment-825169306

**Comment by dougwaldron** [comment]
Feedback from Erin, et al:

> > Mail contacts will be included for all current programs (Emission Inventory, Permitting, etc.), but we would need to discuss it with each program before including the option of electronic communications. 
> 
> We have confirmed with Karen based on a previous discussion that no other Programs/Units in the Air Branch need the e-notice feature for communication. Only Finance will use it for fee notifications.
> 
> We would like to request that the telephone number field also be required. It is not uncommon for mail to be returned and for the email address to be defunct. Having a telephone number will help us contact users, as needed, in Finance, and it may also help Compliance if the facility is ever turned over for fee issues, etc.
> 
> > users can routinely (annually?) be shown a reminder page asking them to review & confirm their current settings 
> 
> Yes, we would like a reminder page for each user to re-confirm communication preferences annually when the system is opened for each new fee year period.
> 
> Based on the new set-up, will we still be able to query or find an existing query to look at:
> * Mailout list for those who have opted in to hardcopy mail only (or not opted in at all)
> * Users who have verified contact information versus users who have not (We could potentially reach out to unverified users to ensure they complete the process)
> * Users who have selected communication preferences versus users who have not selected anything at all (Again, we could potentially reach out to users to encourage them to make a selection)
> 
> Having the ability to query the information or even if it was one large data dump we could then filter or insert a pivot table into would be great for the above items.
> 
> Is there any possibility for the system to send “reminder” emails to those users who have not begun to generate invoice, those who are still “in progress,” and those who have not paid?
> I think we talked about this being a possibility previously, but we wanted to see if it may still be doable? It will help us determine how manual our collections process and reminders will be.

**Comment by dougwaldron** [comment]
#  Review of other contact tables used by fee system

There are three tables with contact info used by the fees system:

* `APBContactInformation`
* `FS_CONTACTINFO`
* `FS_MAILOUT`

<details>
<summary><b>Description of table usage and identification of further action needed</b></summary>

## `APBContactInformation`

This table is used by all APB programs. Data comes from multiple sources, such as permit applications, emissions inventory submittals, etc. Data is editable by staff in the IAIP Edit Contacts form. Updates in GECO should also update this table. 

New contacts added to this table could be added to `Geco_MailContact` as unconfirmed if a contact in that category doesn't exist yet.

Contacts with a key of "40" are fees contacts and can be updated in the GECO Annual Fees submittal. Updating in GECO also writes a snapshot to `FS_CONTACTINFO`. This should be changed to also update `Geco_MailContact`.

### Actions needed

* Remove `FeesContact` view from "Run balance report", drop view, and replace Crystal Report with DataGridView.
* Update `APBContactInformation` when `Geco_MailContact` is updated in GECO.
* Add contact to `Geco_MailContact` as unconfirmed when `APBContactInformation` is updated if address doesn't exist?

## `FS_MAILOUT`

This table holds a snapshot of facility data for each fee year. It holds more than just contact info.

It's populated annually on demand from the IAIP Fee Management tools using data from `APBContactInformation`. Need to retain tools to populate this table, but improve the source of contact data. Could add some notes to indicate where data comes from. Two buttons - "Generate Initial Mailout List" and "Update Mailout List With Current Fee Contacts" - need to be updated.

Mailout info can be updated by staff in the IAIP Fee Audit. Need to update the "Replace mailout contact with current fee contact" button.

Mailout contact info is displayed in the IAIP Fee Audit, Fees Management tools when viewing enrolled facilities or mailout list, and Fee Statistics when viewing "fee universe" details and "not reported" details.

### Actions needed

* Fix data source for `FS_MAILOUT` population.

## `FS_CONTACTINFO`

This table is only written to by GECO users during the annual fee submittal process. User is asked to verify/update contact info (which is pre-populated from either `FS_CONTACTINFO` or `APBContactInformation`). Updating the Fee Contact is one of the steps (step 5) in the Admin Fee Status, and the `PD_FEE_STATUS` stored procedure uses this table when updating fee status. Need to fix the data source for population.

Contact info is displayed on the GECO fee history page, the IAIP Fee Statistics Facility Summary Report (via the `VW_FACILITY_FEE` view), the IAIP Fee Audit Submittal Contact tab, and the Fee Statistics "details" lists.

### Actions needed

* Fix data source for GECO Annual Fees form population, use a stored procedure when saving, and update `Geco_MailContact` at the same time.

</details>

**Comment by dougwaldron** [comment]
<details>
<summary>Where is <code>APBCONTACTINFORMATION</code> updated?</summary>

```
16 results - 7 files

airbranch-db • DB Updates\GECO-485-contact-phone\table-modifications.sql:
  38: update APBCONTACTINFORMATION

iaip • IAIP\Facility\Sub Forms\IAIPEditContacts.vb:
  336:                     Dim query As String = "UPDATE APBCONTACTINFORMATION " &
  395:                         Dim query As String = "UPDATE APBCONTACTINFORMATION " &
  487:                             query = "Update APBContactInformation set " &
  500:                             query = "INSERT INTO APBCONTACTINFORMATION " &
  544:                             query = "INSERT INTO APBCONTACTINFORMATION " &

iaip • IAIP\Facility\Sub Forms\IAIPFacilityCreator.vb:
   615:             SQL = "insert into APBContactInformation " &
  1968:             SQL = "update APBContactInformation set " &

iaip • IAIP\SSPP\SSPPApplicationTrackingLog.vb:
  5784:                         query = "Insert into APBContactInformation " &
  5831:                             query = "Update APBContactInformation set " &
  5871:                             query = "Insert into APBContactInformation " &
  5908:                     query = "Update APBContactInformation set " &

iaip • IAIP\SSPP\Sub Forms\SSPPFeeContact.vb:
  201:             SQL = "Update APBContactInformation set " &

geco • GECO\AnnualFees\Default.aspx.vb:
  590:         Dim Sql3 As String = "Update apbcontactinformation set " &

geco • GECO\ES\esform.aspx.vb:
  909:             query = "Update apbContactInformation Set " &
  927:             query = "Insert into apbContactInformation (" &
```
</details>

**Comment by dougwaldron** [comment]
<details>
<summary>Feedback from 2nd UAT</summary>

# E-Notification Testing Session (8/26/2021)

1. When selecting communication preferences, we completed various scenarios selecting each individual communication preference type. When selecting “Both” email and physical mail for Facility 021-00209, we did not see the facility user information show up under the GECO Contacts tab in the Facility Summary. 

2. After the user makes a selection for communication preferences and is taken to the below “Communication Preferences” section, we would like to request the following updates:

    1. The email address field must be mandatory/required for those users selecting email only or both email and physical mail. We were able to bypass this page during testing even when we selected email communication as a preference, which makes it easy for a user to select a communication preference and never enter or verify an email address.

    2. Any user should not be able to navigate past this page without at least clicking “Save” at the bottom. This ensures the user has reviewed any listed physical address for accuracy (when applicable) and/or that the user has entered and saved an email address.

3. On the mailout lists (screen shot below), we would like to request that the “Email Address” column be added back to the table regardless of if the customer is opted in to physical mail only, email only, or both. This is helpful information should we ever need to try to contact multiple users via email for marketing purposes, such as encouraging users to opt in to electronic communication in the future.

4. While we have discussed this briefly in the past, we wanted to ask again about the possibility of sending e-notifications directly from IAIP. Meaning, we would have e-notification templates within the system or ways to upload templates, and we could select specific facilities to receive those communications. Given the work that has been done so far, is this still a possibility, or would this be disastrous? 

5. On the “Fee Summary” table (screen shot below), can the “Communication Preferences” column be added for each facility to help us identify facilities who need to be sent reminders regarding fee payments, etc.? Since this table is the area of the system we usually check to identify those who have not reported/generated an invoice as well as those who have not paid, knowing if they are opted in to electronic communication versus mail would be helpful.
</details>

**Comment by dougwaldron** [comment]
Published to prod from efd11847
